mkosi/__init__.py

```diff
@@ -543,6 +551,184 @@ def install_systemd_boot(state: MkosiState) -> None:
+        if (p := state.root / d).exists() and any(p.iterdir()):
+            return p
+
+    return None
+
+
+def find_grub_install(state: MkosiState) -> Optional[Path]:
+    path = ":".join(os.fspath(p) for p in [state.root / "usr/bin", state.root / "usr/sbin"])
+
+    install = shutil.which("grub2-install", path=path) or shutil.which("grub-install", path=path)
+    if not install:
+        return None
+
+    return Path(install).relative_to(state.root)
+
+
+def want_grub_bios(state: MkosiState, partitions: Sequence[Partition] = ()) -> bool:
+    if state.config.bootable == ConfigFeature.disabled:
+        return False
+
+    if state.config.output_format != OutputFormat.disk:
+        return False
+
+    if state.config.bios_bootloader != BiosBootloader.grub:
+        return False
+
+    have = find_grub_bios_directory(state) is not None
+    if not have and state.config.bootable == ConfigFeature.enabled:
+        die("A BIOS bootable image with grub was requested but grub for BIOS is not installed")
+
+    bios = any(p.type == Partition.GRUB_BOOT_PARTITION_UUID for p in partitions)
+    if partitions and not bios and state.config.bootable == ConfigFeature.enabled:
+        die("A BIOS bootable image with grub was requested but no BIOS Boot Partition was configured")
+
+    root = any(p.type.startswith("root") or p.type.startswith("usr") for p in partitions)
+    if partitions and not root and state.config.bootable == ConfigFeature.enabled:
+        die("A BIOS bootable image with grub was requested but no root or usr partition was configured")
+
+    install = find_grub_install(state) is not None
+    if not install and state.config.bootable == ConfigFeature.enabled:
+        die("A BIOS bootable image with grub was requested but grub-install was not found")
+
+    want = (have and bios and root and install) if partitions else have
+
+    control = Path("/dev/loop-control")
+    if not control.exists() or control.stat().st_uid != os.getuid():
+        if state.config.bootable == ConfigFeature.enabled:
+            if not control.exists():
+                die("A BIOS bootable image with grub was requested but grub cannot be installed because /dev/loop-control is not available")
+            else:
+                die(
+                    textwrap.dedent(
+                        """\
+                        A BIOS bootable image with grub was requested but grub cannot be installed because
+                        mkosi is running in a user namespace. Run mkosi with root privileges outside of any
+                        unprivileged containers to solve this problem.
+                        """
+                    )
+                )
+        if want:
+            if not control.exists():
+                logging.warning("Cannot install grub because /dev/loop-control is missing, skipping")
+            else:
+                logging.warning("Cannot install grub because mkosi is running in a user namespace, skipping")
+        return False
```

Comment:
I think this can be reordered like this to save one level of logic
```suggestion
    if not control.exists():
        if state.config.bootable == ConfigFeature.enabled:
            die("A BIOS bootable image with grub was requested but grub cannot be installed because /dev/loop-control is not available")
        elif want:
            logging.warning("Cannot install grub because /dev/loop-control is missing, skipping")
        return False
    elif control.stat().st_uid != os.getuid():
        if state.config.bootable == ConfigFeature.enabled:
            die(
                textwrap.dedent(
                    """\
                    A BIOS bootable image with grub was requested but grub cannot be installed because
                    mkosi is running in a user namespace. Run mkosi with root privileges outside of any
                    unprivileged containers to solve this problem.
                    """
                )
            )
        elif want:
            logging.warning("Cannot install grub because mkosi is running in a user namespace, skipping")
        return False
```